File: 230-kth-smallest-element-in-a-bst/leochang/230.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
	int kthSmallest(TreeNode* root, int k) {
		return InOrderTraverse(root, k);
	}
private:
	int InOrderTraverse(TreeNode* root, int &k) {
		if (root == NULL) return -1;
		int ans = InOrderTraverse(root->left, k);
		k--;
		if (k == 0) return root->val;
		if (ans == -1) ans = InOrderTraverse(root->right, k);
		return ans;
	}
};

File: 230-kth-smallest-element-in-a-bst/ken/230.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isFound(self):
        return self.cnt == self.k

    def rec(self, root):
        if root.left != None:
            val = self.rec(root.left)
            if self.isFound():
                return val

        self.cnt += 1
        if self.isFound():
            return root.val

        if root.right != None:
            val = self.rec(root.right)
            if self.isFound():
                return val

    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """

        self.cnt = 0
        self.k = k

        return self.rec(root)